### PR TITLE
Add a socket interface for cyberVirtua/dexpo/issues/3

### LINKED ISSUE #3: Daemon that will screenshot desktops and expose them through socket interface
The daemon has to
- Make and store Pixmaps of each virtual desktop together with the _NET_DESKTOP_NAME
- Send Pixmaps to GUI client over unix sockets
  - Make a socket server (daemon) and a socket client (GUI)

### PR DESCRIPTION
**Pull request by mmskv**
This interface is used to send and receive pixmaps.
Currently implementation is really ugly and needs to be changed. Suggestions are welcome.
Add simple tests to check implementation. CMakeLists.txt are ugly and need to be refactored.

### CODE REVIEW

File: src/dexpo_socket.hpp [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+public:
+  int fd;       ///< Socket File Descriptor
+  bool running; ///< Thread status
+
+  dexpo_socket ();
+  ~dexpo_socket ();
+
+  void send (dexpo_pixmap &) const;
+  void send (std::vector<dexpo_pixmap>) const;
+
```
Comment (TheHussky) [inline src/dexpo_socket.hpp]: Why vector is used? Isn't just an array of pixmaps more fitting?

File: src/dexpo_socket.cpp [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+  this->fd = socket (AF_UNIX, SOCK_STREAM, 0);
+  handle_error (this->fd);
+
+  // Trick compiler into thinking that we pass pointer to sockaddr struct
+  // https://stackoverflow.com/questions/21099041
+  auto *sock_addr = reinterpret_cast<struct sockaddr *> (&sock_name);
+
+  // Connect to named socket
+  int e = connect (this->fd, sock_addr, sizeof (sock_name));
+  if (e == -1) // Socket doesn't exist
```
Comment (TheHussky) [inline src/dexpo_socket.cpp]: Why not add it to handle_error?

File: src/dexpo_socket.hpp (hunk 1 above)
Comment (mmskv) [reply src/dexpo_socket.hpp]: vector is the c++ style of handling arrays. and it is RAII compliant

File: src/dexpo_socket.cpp [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+
+/**
+ * Request and receive dexpo_pixmaps from daemon
+ */
+std::vector<dexpo_pixmap>
+dexpo_socket::get_pixmaps () const
+{
+  // Send pixmap request
+  char cmd = kRequestPixmaps;
+  if (write (this->fd, &cmd, 1) == -1)
```
Comment (TheHussky) [inline src/dexpo_socket.cpp]: handle errors?

File: src/dexpo_socket.cpp (hunk 2 above)
Comment (mmskv) [reply src/dexpo_socket.cpp]: handle_error will `throw`, but here we need to handle the error differently
I need to add custom error types and improve error handling in general

File: src/dexpo_socket.cpp [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+      throw;
+    };
+
+  dexpo_pixmap pixmap; // Pixmap to copy data in
+  std::vector<dexpo_pixmap> pixmap_array;
+
+  size_t num;
+  auto rcv_bytes = read (this->fd, &num, sizeof (num));
+
+  if (rcv_bytes == -1)
```
Comment (TheHussky) [inline src/dexpo_socket.cpp]: handle_errors?

File: src/dexpo_socket.cpp (hunk 3 above)
Comment (mmskv) [reply src/dexpo_socket.cpp]: If write fails, it means that it got SIGPIPE. I don't think there is a way out of it

File: src/dexpo_socket.cpp [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+      pixmap_array.push_back (pixmap);
+    }
+  return pixmap_array;
+};
+
+void
+dexpo_socket::send_pixmaps_on_event (daemon_event e,
+                                     std::vector<dexpo_pixmap> &pixmap_array)
+{
+  auto data_fd = accept (this->fd, NULL, NULL);
```
Comment (TheHussky) [inline src/dexpo_socket.cpp]: What do NULLs stand for?

File: src/dexpo_socket.cpp (hunk 4 above)
Comment (mmskv) [reply src/dexpo_socket.cpp]: It's on my TODO list, no need to mention it everywhere

File: src/dexpo_socket.cpp (hunk 5 above)
Comment (mmskv) [reply src/dexpo_socket.cpp]: Anonymous socket

File: src/dexpo_socket.cpp [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+  char cmd = -1;
+  this->running = true;
+  while (this->running)
```
Comment (TheHussky) [inline src/dexpo_socket.cpp]: Is this generally ok? Is running state always true? How does it respond to errors?

Comment (mmskv) [reply src/dexpo_socket.cpp]: This is how I stop it
https://github.com/cyberVirtua/dexpo/blob/sockets/test/socket.cpp#L57
Couldn't come up with a better way. What do you think?

Comment (TheHussky) [reply src/dexpo_socket.cpp]: The whole boolean status idea is fine imho. Maybe add running=false to default destructor? Or add an event type that'll make it false?

Comment (mmskv) [reply src/dexpo_socket.cpp]: > Maybe add running=false to default destructor

sounds great

Comment (mmskv) [reply src/dexpo_socket.cpp]: or does it?
the objects are created in the order: `dexpo_pixmap` -> `thread`
so thread destructor is called before `dexpo_pixmap`'s

File: src/dexpo_socket.hpp [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+  std::vector<dexpo_pixmap> get_pixmaps () const;
+  void send_pixmaps_on_event (std::vector<dexpo_pixmap> &);
```
Comment (mmskv) [inline src/dexpo_socket.hpp]: add `const` to args

File: src/dexpo_socket.cpp [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+ * and sends pixmaps one by one in return
+ */
+void
+dexpo_socket::send_pixmaps_on_event (std::vector<dexpo_pixmap> &pixmap_array)
+{
+  auto data_fd = accept (this->fd, NULL, NULL); // Anonymous socket
+  char cmd = -1;
+
+  this->running = true; // Can later turn off the loop
+  while (this->running) // FIXME This loop exit is ugly
```
Comment (mmskv) [inline src/dexpo_socket.cpp:135]: make `this->running` atomic to avoid race condition

File: src/dexpo_socket.cpp [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+        {
+          // First write -- number of subsequent packets
+          auto num = pixmap_array.size ();
+          auto s = write (data_fd, &num, sizeof (num));
+          is<write_error> (s);
+
+          // Writes from second to `num+1` -- subsequent packets
+          for (auto &pixmap : pixmap_array)
+            {
+              s = write (data_fd, &pixmap, sizeof (pixmap));
```
Comment (mmskv) [inline src/dexpo_socket.cpp:151]: make `pixmap_array` mutex to avoid race condition

**Review by mmskv — COMMENTED** [review]
Race conditions are bad

**Pull request merged**